src/bst.py

```diff
@@ -0,0 +1,285 @@
+        of a binary tree(left child, parent, right child), one value at a time.
+        """
+        try:
+            for node in self.left._in_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        yield self.val
+        try:
+            for node in self.right._in_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+
+    def _pre_order(self):
+        """
+        This internal method is a generator that will output preorder traversal
+        of a binary tree(parent, left child, right child), one value at a time.
+        """
+        yield self.val
+        try:
+            for node in self.left._pre_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        try:
+            for node in self.right._pre_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+
+    def _post_order(self):
+        """
+        This internal method is a generator that will output postorder traversal
+        of a binary tree(left child, right child, parent), one value at a time.
+        """
+        try:
+            for node in self.left._post_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        try:
+            for node in self.right._post_order():
+                if node is None:
+                    pass
+                else:
+                    yield node
+        except AttributeError:
+            pass
+        yield self.val
+
+
+class BinarySearchTree(object):
+    """BinarySearchTree implements a Binary Search Tree data structure
+    and associated methods."""
+
+    def __init__(self, root=None):
+        """Create an instance of our Binary Search Tree w/ supplied
+        input, or an emptry tree."""
+        self.length = 0
+        if root:
+            self.root = Node(root)
+            self.length = 1
+        else:
+            self.root = root
+
+    def insert(self, val, data=None):
+        """Inserts a value  into the BST.  If the value is already in
+        the BST it will be ingored."""
+        if not self.contains(val):
+            if self.root:
+                self.root._insert(val, data)
+            else:
+                self.root = Node(val, data)
+            self.length += 1
+
+    def contains(self, val):
+        """Will return True if val is in the BST, or False if it's not."""
+        try:
+            return self._contains(val, self.root)
+        except AttributeError:
```

Comment:
Where are you expecting the `AttributeError` to pop up? Doesn't look it can happen inside this function. If you're expecting an AttributeError to occur in `self._contains()`, your `try except` block should also be in `self._contains()`.